hudi-spark/src/main/scala/org/apache/hudi/AvroConversionUtils.scala

```diff
@@ -78,4 +79,21 @@ object AvroConversionUtils {
   def convertAvroSchemaToStructType(avroSchema: Schema): StructType = {
     SchemaConverters.toSqlType(avroSchema).dataType.asInstanceOf[StructType]
   }
+
+  private def deserializeRow(encoder: ExpressionEncoder[Row], internalRow: InternalRow): Row = {
+    // First attempt to use spark2 API for deserialization, otherwise attempt with spark3 API
+    try {
+      val spark2method = encoder.getClass.getMethods.filter(method => method.getName.equals("fromRow")).last
+      spark2method.invoke(encoder, internalRow).asInstanceOf[Row]
+    } catch {
+      case e: NoSuchElementException => spark3Deserialize(encoder, internalRow)
```

Comment:
how expensive would this be in practice? any thoughts?

Reply:
wondering if we can use `SparkContext.version` or something to figure it out upfront..

Reply:
I think checking spark version would be a good idea to prevent the failed call everytime, let me look into it. I don't believe the reflection performance is significant as a whole, especially if we can figure out the spark version upfront. We do something similar here:
https://github.com/apache/hudi/pull/1638/commits/9f1284374e72717222c51ea681dc2a5ceb696a50

Reply:
I was concerned about the exception thrown each time.. in the case you mentioned, its obtaining the description once.. this would be in the fast path.. Anyways.. lets hope the version thing is feasible..

Reply:
I think `org.apache.spark.SPARK_VERSION` could help us out here. I've implemented it in this commit: https://github.com/sbernauer/hudi/commit/a4f1866f5be56639958479e9a597ae8c4d3d8f4f. But I noticed that it had a huge performance impact (I assume due to the reflection)

Reply:
I would suggest to change the method to something like this
```
import org.apache.spark.SPARK_VERSION

  private def deserializeRow(encoder: ExpressionEncoder[Row], internalRow: InternalRow): Row = {
    // TODO remove reflection if Spark 2.x support is dropped
    if (SPARK_VERSION.startsWith("2.")) {
      val spark2method = encoder.getClass.getMethod("fromRow", classOf[InternalRow])
      spark2method.invoke(encoder, internalRow).asInstanceOf[Row]
    } else {
      val deserializer = encoder.getClass.getMethod("createDeserializer").invoke(encoder)
      val aboveSpark2method = deserializer.getClass.getMethod("apply", classOf[InternalRow])
      aboveSpark2method.invoke(deserializer, internalRow).asInstanceOf[Row]
    }
  }
```

Reply:
Thanks for the suggestion! I will try making the same change as that commit you linked but I am going to see if I can test the performance impact first.

Reply:
The commit with all the patches (including better reflection) is this here: https://github.com/sbernauer/hudi/commit/0dd7172521a92d9a13425dde4a23c565cb37a533

Reply:
@nsivabalan are you able to take a swing at this for 0.6.0? this would be good to have

Reply:
@sbernauer 's check is better I think. 

On the reflection itself, @bschell would it help if we cached the method objects, instead of recreating everytime. have you attempted anything like that?